update: Refactoring Code-class and new Custom pygments based formatter

OliverStrait opened the pull request:
## Overview: What does this pull request change?
* Refactoring `Code` class
*  Added new Formatter class `CodeColorFormatter`
*  Added testes and new utility-file to test reading and formating from code file.

<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->

<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
* Old Code parser is very convoluted and hard to understand  code of custom python parser with poor naming, and jumpping and overlapping procedure. It used `pygments` HtmlFormatter as backend and after parsed html to `list[tuple[str,str]]` structure which is used for coloring.
* New custom formatter skips html and transforms code directly to list-mapping using `pygments` lexers and tokenizer functions.
* Better naming and capsulating responsibility to a logical chuncks.
* New system is just about 2 to 3 times faster, but Mobject constructors are still performance bottlenecks

![code_class-new_perf](https://github.com/user-attachments/assets/016363b0-7758-4957-8fb0-ec0821c1dfc8)

<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->

## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->

## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->

* If Paragraph could support color to text mapping we could remove whole block of re-coloring code after object-construction (little performance boost).
* Tested with pygments 2.18.0 and python language
** There may be some odd cases when pygments Tokenizer sneak newlines to unexpected places. Testes with different syntax and languages would be good to perform.
** If uncatched newlines is passed to `Paragraph` inside of string literals  it will broke re-coloring an line-numbers.

 
<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

JasonGrace2282 commented:
This seems very promising, thanks for your work! Hopefully, I'll get the chance to look at this in a bit :)

OliverStrait commented:
* I made parser more general to handle odd cases where tokenizer produce string literals with newlines hidden inside.
* Somehow Paragraph class cannot handle empty newlines at end of string (therefore line numbers will missalign)  so I added parser to remove those.
* (also fixed messed up git history)

JasonGrace2282 commented:
I'll close this as we now have #4115
